pkg/testworkflows/testworkflowexecutor/prepared.go

```diff
@@ -0,0 +1,347 @@
+	e.execution.Id = primitive.NewObjectIDFromTimestamp(time.Now()).Hex()
+	return e
+}
+
+func (e *IntermediateExecution) ID() string {
+	return e.execution.Id
+}
+
+func (e *IntermediateExecution) GroupID() string {
+	return e.execution.GroupId
+}
+
+func (e *IntermediateExecution) SetName(name string) *IntermediateExecution {
+	e.execution.Name = name
+	return e
+}
+
+func (e *IntermediateExecution) SetGroupID(groupID string) *IntermediateExecution {
+	if groupID == "" {
+		e.execution.GroupId = e.execution.Id
+	} else {
+		e.execution.GroupId = groupID
+	}
+	return e
+}
+
+func (e *IntermediateExecution) AppendTags(tags map[string]string) *IntermediateExecution {
+	e.dirty = true
+	maps.Copy(e.tags, tags)
+	return e
+}
+
+func (e *IntermediateExecution) SetKubernetesObjectName(name string) *IntermediateExecution {
+	e.execution.TestWorkflowExecutionName = name
+	return e
+}
+
+func (e *IntermediateExecution) SetDisabledWebhooks(disabled bool) *IntermediateExecution {
+	e.execution.DisableWebhooks = disabled
+	return e
+}
+
+func (e *IntermediateExecution) SetRunningContext(runningContext *testkube.TestWorkflowRunningContext) *IntermediateExecution {
+	e.execution.RunningContext = runningContext
+	return e
+}
+
+func (e *IntermediateExecution) WorkflowName() string {
+	if e.cr == nil {
+		panic("workflow not set yet")
+	}
+	return e.cr.Name
+}
+
+func (e *IntermediateExecution) Name() string {
+	return e.execution.Name
+}
+
+func (e *IntermediateExecution) Execution() *testkube.TestWorkflowExecution {
+	if e.cr == nil {
+		panic("workflow not set yet")
+	}
+	if e.dirty {
+		e.dirty = false
+		e.execution.ResolvedWorkflow = testworkflowmappers.MapKubeToAPI(e.cr)
+		e.execution.Tags = make(map[string]string)
+		if e.cr.Spec.Execution != nil {
+			// TODO: Should resolve the expressions? (`{{"{{"}}` becomes `{{`)
+			maps.Copy(e.execution.Tags, e.cr.Spec.Execution.Tags)
+			maps.Copy(e.execution.Tags, e.tags)
+		}
+	}
+	return e.execution
+}
+
+func (e *IntermediateExecution) PrependTemplate(name string) *IntermediateExecution {
+	if name == "" {
+		return e
+	}
+	if e.cr == nil {
+		e.prepended = append(e.prepended, name)
+		return e
+	}
+	e.dirty = true
+	testworkflowresolver.AddGlobalTemplateRef(e.cr, testworkflowsv1.TemplateRef{
+		Name: testworkflowresolver.GetDisplayTemplateName(name),
+	})
+	return e
+}
+
+func (e *IntermediateExecution) SensitiveData() map[string]string {
+	return e.sensitiveData.Data
+}
+
+func (e *IntermediateExecution) simplifyCr() error {
+	crMachine := testworkflowconfig.CreateWorkflowMachine(&testworkflowconfig.WorkflowConfig{Name: e.cr.Name, Labels: e.cr.Labels})
+	err1 := expressions.Simplify(e.cr, crMachine)
+	err2 := expressions.SimplifyForce(&e.sensitiveData.Data, crMachine)
+	err := errors2.Join(err1, err2)
+	e.dirty = true
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *IntermediateExecution) ApplyDynamicConfig(config map[string]string) error {
+	if e.cr == nil {
+		panic("workflow not set yet")
```

Comment:
should we panic or return an error?

Reply:
there and below

Reply:
I think that we should panic - if something will try to call these methods before setting the workflow, it means that the code that is using this model is broken.

On the other hand, handling it gracefully should work in most of the cases (as long as the code that is using it has error handling), so I'm not sure.

Do you have some preference? Either way is fine for me - both have advantages and disadvantages.

Reply:
if we return error, I would not panic. still it's up to you to decide

Reply:
I have just noticed, that we have it also in other places in this file, but where it cannot be handled gracefully - I think that I'll just leave it there for consistency